cl/corpus_importer/management/commands/financial_disclosures.py

```diff
@@ -0,0 +1,344 @@
+        ):
+            logger.info("Could not extract data from this document")
+            return {}
+
+    logger.info("Processing extracted data")
+    return extractor_response.json()
+
+
+def get_report_type(extracted_data: dict) -> int:
+    """Get report type if available
+
+    :param extracted_data: Document information
+    :return: Disclosure type
+    """
+    if extracted_data.get("initial"):
+        return FinancialDisclosure.INITIAL
+    elif extracted_data.get("nomination"):
+        return FinancialDisclosure.NOMINATION
+    elif extracted_data.get("annual"):
+        return FinancialDisclosure.ANNUAL
+    elif extracted_data.get("final"):
+        return FinancialDisclosure.FINAL
+
+
+def save_disclosure(
+    extracted_data: dict, disclosure: FinancialDisclosure
+) -> None:
+    """Save financial data to system.
+
+    Wrapped in a transaction, we fail if anything fails.
+
+    :param disclosure: Financial disclosure
+    :param extracted_data: disclosure
+    :return:None
+    """
+    addendum = "Additional Information or Explanations"
+
+    # Process and save our data into the system.
+    with transaction.atomic():
+        disclosure.has_been_extracted = True
+        disclosure.addendum_content_raw = extracted_data[addendum]["text"]
+        disclosure.addendum_redacted = extracted_data[addendum]["is_redacted"]
+        disclosure.is_amended = extracted_data.get("amended") or False
+        disclosure.report_type = get_report_type(extracted_data)
+        disclosure.save()
+
+        for investment in extracted_data["sections"]["Investments and Trusts"][
+            "rows"
+        ]:
+            Investment.objects.create(
+                financial_disclosure=disclosure,
+                redacted=True
+                in [v["is_redacted"] for _, v in investment.items()],
+                description=investment["A"]["text"],
+                has_inferred_values=investment["A"]["inferred_value"],
+                income_during_reporting_period_code=investment["B1"]["text"],
+                income_during_reporting_period_type=investment["B2"]["text"],
+                gross_value_code=investment["C1"]["text"],
+                gross_value_method=investment["C2"]["text"],
+                transaction_during_reporting_period=investment["D1"]["text"],
+                transaction_date_raw=investment["D2"]["text"],
+                transaction_value_code=investment["D3"]["text"],
+                transaction_gain_code=investment["D4"]["text"],
+                transaction_partner=investment["D5"]["text"],
+            )
+
+        for agreement in extracted_data["sections"]["Agreements"]["rows"]:
+            Agreements.objects.create(
+                financial_disclosure=disclosure,
+                redacted=True
+                in [v["is_redacted"] for _, v in agreement.items()],
+                date=agreement["Date"]["text"],
+                parties_and_terms=agreement["Parties and Terms"]["text"],
+            )
```

Comment:
Here and in the other simple for-loops are good opportunities to use django's `bulk_create` pattern.You'd do fewer queries on the DB and wind up with something like:

```python3
Agreements.objects.bulk_create([
    Agreement(
        financial_disclosure=disclosure,
        redacted=any(v["is_redacted"] for v in agreement.values()),
        date=agreement["Date"]["text"],
        parties_and_terms=agreement["Parties and Terms"]["text"],
    ) for agreement in extracted_data["sections"]["Agreements"]["rows"],
])
```

The real reason for this is to be friendly to the DB when you're doing a lot of insertions, and I think we're probably doing enough here where it's worth a quick refactor. I imagine in some cases it'd reduce things from > 100 queries down to five or so.

I think that's a bit trickier to read at first, but once you've seen it a few times it becomes easier because the list comprehension (sort of) guarantees that there's minimal tricky logic in the loop — Whereas a normal for loop can do anything, list comprehensions are usually pretty dumbed down (unless you're crazy).